### PR TITLE
Separate registration for global ComWrappers instance for tracker support and marshalling

### LINKED ISSUE #35530: ComWrappers API does not provide a mechanism to identify when it is called from old-style APIs
In #33485, we wired up the ComWrappers API to be called first-chance from the Marshal APIs and from the IL stub marshalers. This has started to prove a problem for CsWinRT, since there's no way to determine when to use the ComWrappers' implementation and when to use the runtime's built-in APIs.

In particular, the CsWinRT ComWrappers implementation of `ComputeVtables` has to be able to handle all types, including types that do not implement any WinRT APIs, to correctly implement WinRT semantics for WinUI. Additionally, since WinUI depends on the reference tracking support from ComWrappers, the CsWinRT ComWrappers implementation has to be registered as the global implementation. As a result, if someone such as WPF calls `Marshal.GetIUnknownForObject`, this call will go down the ComWrappers route first if the Windows SDK is also being used. WPF is expecting to get the built-in runtime-implemented CCWs, but by default CsWinRT would provide one instead.

This proves to be a problem when the user (either WPF or someone else directly using the Marshal APIs or using COM objects in P/Invoke signatures) is trying to pass a managed implementations of COM APIs to native. CsWinRT's ComWrappers only support the WinRT model (as it should be, they shouldn't have to completely re-implement the runtime's CCW/RCW implementation), so they would create a WinRT CCW of the managed object.

We could attempt to use a heuristic in the CsWinRT ComWrappers implementation to determine whether to use the ComWrappers or just return null and fall back, but that heuristic may be tricky or downright impossible to get exactly right. I've included an example below where the heuristic is extremely difficult if not impossible:

Let's take a class `class Foo : System.Collections.IEnumerable { ... }` and a P/Invoke signature `void Bar(System.Collections.IEnumerable ienum)`. In the WinRT world, we would want to treat this as a `Microsoft.UI.Xaml.Interop.IBindableIterable`. However, the built-in CCW system would generate an `IDispatch` implementation that exposes the `GetEnumerator` method as `DISPID_NEWENUM`.

Since there is no way to determine from within a `ComWrappers` implementation if it is being called from the `Marshal` APIs or a P/Invoke marshaler (without attempting to just look up the call stack), the CsWinRT ComWrappers would never be able to know if it should marshal an instance of `Foo` itself or if it should let the runtime marshal it.

There is a corresponding problem for the RCW direction, but using a heuristic of "does this object implement IInspectable" works well as a heuristic for ComWrappers.

There are a few possible solutions here:
- Provide an additional flag on the `CreateObjectFlags` and `CreateComInterfaceFlags` enumeration to denote when the ComWrappers methods are being called from the `Marshal` APIs or from an IL stub.
- Remove the support wiring up a ComWrappers instance as a first-chance handler for the Marshal APIs or from an IL stub.

I'm also open to additional ideas.

cc: @AaronRobinsonMSFT @elinor-fung @jkotas

### PR DESCRIPTION
**Pull request by elinor-fung**
Keep registered global instance of `ComWrappers` for reference tracker support separate from global instance for marshalling.

On `ComWrappers`, replace
```C#
public void RegisterAsGlobalInstance();
```
with two separate methods for registration:
```C#
public static void RegisterForTrackerSupport(ComWrappers instance);
public static void RegisterForMarshalling(ComWrappers instance);
```

Resolves #35530 

cc @AaronRobinsonMSFT @jkoritzinsky @jkotas

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs [hunk 1]
```diff
@@ -164,10 +179,23 @@ private static bool TryGetOrCreateComInterfaceForObjectInternal(ComWrappers? imp
         /// </remarks>
         protected unsafe abstract ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count);
 
-        // Call to execute the abstract instance function
-        internal static unsafe void* CallComputeVtables(ComWrappers? comWrappersImpl, object obj, CreateComInterfaceFlags flags, out int count)
+        // Called by the runtime to execute the abstract instance function
+        internal static unsafe void* CallComputeVtables(ComWrappers? comWrappersImpl, object obj, CreateComInterfaceFlags flags, ComWrappersScenario scenario, out int count)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:183]: Lets put the `ComWrappersScenario scenario` as the first argument in the call.

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs [hunk 2]
```diff
@@ -203,10 +231,23 @@ public object GetOrCreateObjectForComInstance(IntPtr externalComObject, CreateOb
         /// </remarks>
         protected abstract object? CreateObject(IntPtr externalComObject, CreateObjectFlags flags);
 
-        // Call to execute the abstract instance function
-        internal static object? CallCreateObject(ComWrappers? comWrappersImpl, IntPtr externalComObject, CreateObjectFlags flags)
+        // Called by the runtime to execute the abstract instance function.
+        internal static object? CallCreateObject(ComWrappers? comWrappersImpl, IntPtr externalComObject, CreateObjectFlags flags, ComWrappersScenario scenario)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:235]: Put `ComWrappersScenario scenario` as first argument.

File: src/coreclr/src/vm/interoplibinterface.cpp [hunk 3]
```diff
@@ -406,10 +406,20 @@ namespace
     // Defined handle types for the specific object uses.
     const HandleType InstanceHandleType{ HNDTYPE_STRONG };
 
+    // Scenarios for ComWrappers usage.
+    // These values should match the managed definition in ComWrappers.
+    enum ComWrappersScenario
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/vm/interoplibinterface.cpp:411]: `enum class`

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs [hunk 4]
```diff
@@ -268,32 +309,59 @@ private static bool TryGetOrCreateObjectForComInstanceInternal(ComWrappers? impl
...
+            if (instance == null)
+                throw new ArgumentNullException(nameof(instance));
+
+            if (null != Interlocked.CompareExchange(ref s_globalInstanceForMarshalling, instance, null))
             {
                 throw new InvalidOperationException(SR.InvalidOperation_ResetGlobalComWrappersInstance);
             }
 
-            SetGlobalInstanceRegistered();
+            SetGlobalInstanceRegisteredForMarshalling();
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:359]: We should comment why this is needed in this case, but not in the other.

**Review by AaronRobinsonMSFT — APPROVED** [review]
LGTM

**Review by jkotas — APPROVED** [review]

**Pull request merged**